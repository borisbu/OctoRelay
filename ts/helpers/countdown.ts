export const formatDeadline = (time: number): string => {
  let unit: "second" | "minute" | "hour" = "second";
  let timeLeft = (time - Date.now()) / 1000;
  if (timeLeft >= 60) {
    timeLeft /= 60;
    unit = "minute";
  }
  if (timeLeft >= 60) {
    timeLeft /= 60;
    unit = "hour";
  }
  const isLastMinute = unit === "minute" && timeLeft < 2;
  const formattedTimeLeft = new Intl.NumberFormat(LOCALE, {
    style: "unit",
    unitDisplay: "long",
    minimumFractionDigits: isLastMinute ? 1 : 0,
    maximumFractionDigits: isLastMinute ? 1 : 0,
    unit,
  }).format(Math.max(0, timeLeft));
  return `in ${formattedTimeLeft}`;
};

const getCountdownDelay = (deadline: number): number =>
  deadline - Date.now() > 120000 ? 60000 : 1000;

export const setCountdown = (
  selector: JQuery,
  deadline: number
): (() => void) => {
  const delay = getCountdownDelay(deadline);
  let disposer: () => void;
  const interval = setInterval(() => {
    const isVisible = selector.is(":visible");
    if (!isVisible) {
      return disposer();
    }
    selector.text(formatDeadline(deadline));
    const nextDelay = getCountdownDelay(deadline);
    if (nextDelay !== delay) {
      disposer();
      disposer = setCountdown(selector, deadline); // reset with new interval
    }
  }, delay);
  disposer = () => clearInterval(interval);
  return disposer;
};
