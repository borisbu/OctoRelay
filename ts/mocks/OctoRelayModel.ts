export const modelMock = jest.fn();

jest.mock("../model/OctoRelayModel", () => ({ OctoRelayViewModel: modelMock }));
