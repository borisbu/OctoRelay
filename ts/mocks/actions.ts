export const cancelMock = jest.fn();

jest.mock("../helpers/actions", () => ({
  cancelTask: cancelMock,
}));
