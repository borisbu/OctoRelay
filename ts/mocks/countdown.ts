export const disposerMock = jest.fn();
export const countdownMock = jest.fn(() => disposerMock);

jest.mock("../helpers/countdown", () => ({
  setCountdown: countdownMock,
}));
