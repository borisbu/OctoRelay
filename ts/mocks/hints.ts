export const clearMock = jest.fn();
export const showMock = jest.fn();

jest.mock("../helpers/hints", () => ({
  clearHints: clearMock,
  showHints: showMock,
}));
