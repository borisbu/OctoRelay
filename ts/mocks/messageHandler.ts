export const handlerMock = jest.fn();

jest.mock("../model/messageHandler", () => ({
  makeMessageHandler: jest.fn(() => handlerMock),
}));
