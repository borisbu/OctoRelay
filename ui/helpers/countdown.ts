export const formatDeadline = (
  time: number,
  locales = [LOCALE, undefined],
): string => {
  let unit: "second" | "minute" | "hour" = "second";
  let timeLeft = (time - Date.now()) / 1000;
  if (timeLeft >= 60) {
    timeLeft /= 60;
    unit = "minute";
  }
  if (timeLeft >= 60) {
    timeLeft /= 60;
    unit = "hour";
  }
  const isLastMinute = unit === "minute" && timeLeft < 2;
  for (const locale of locales) {
    try {
      const formattedTimeLeft = new Intl.NumberFormat(locale, {
        style: "unit",
        unitDisplay: "long",
        minimumFractionDigits: isLastMinute ? 1 : 0,
        maximumFractionDigits: isLastMinute ? 1 : 0,
        unit,
      }).format(Math.max(0, timeLeft));
      return `in ${formattedTimeLeft}`;
    } catch (error) {
      console.warn(`Failed to format time using ${locale} locale`, error);
    }
  }
  return `in ${timeLeft} ${unit}${timeLeft > 1 ? "s" : ""}`;
};

export const getCountdownDelay = (deadline: number): number =>
  deadline - Date.now() > 120000 ? 60000 : 1000;

export const setCountdown = (
  selector: JQuery,
  deadline: number,
): (() => void) => {
  const delay = getCountdownDelay(deadline);
  let disposer: () => void;
  const interval = setInterval(() => {
    const isVisible = selector.is(":visible");
    if (!isVisible) {
      return disposer();
    }
    selector.text(formatDeadline(deadline));
    const nextDelay = getCountdownDelay(deadline);
    if (nextDelay !== delay) {
      disposer();
      disposer = setCountdown(selector, deadline); // reset with new interval
    }
  }, delay);
  disposer = () => clearInterval(interval);
  return disposer;
};
