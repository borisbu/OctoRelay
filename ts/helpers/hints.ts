import type { Hint, PopoverItem } from "../types/Hints";
import { closeBtnHTML, closeBtnId } from "./const";
import { formatDeadline, setCountdown } from "./countdown";
import { hasUpcomingTask } from "./narrowing";
import { cancelTask } from "./actions";

interface AddPopoverProps {
  target: JQuery;
  title: string;
  content: string[];
  navbar: JQuery;
  items: PopoverItem[];
  originalSubject: string;
}

export const clearHints = (btn: JQuery) =>
  btn.tooltip("destroy").popover("destroy");

const addTooltip = (btn: JQuery, text: string) =>
  btn.tooltip({ placement: "bottom", title: text });

const addPopover = ({
  target,
  title,
  content,
  navbar,
  items,
  originalSubject,
}: AddPopoverProps) => {
  target
    .popover({
      title,
      html: true,
      animation: false,
      placement: "bottom",
      trigger: "manual",
      content: content.join(""),
    })
    .popover("show");
  const closeBtn = navbar.find(`#${closeBtnId}`);
  const countdownDisposers = items.map(
    ({ cancelId, timeTagId, deadline, cancel }) => {
      const cancelBtn = navbar.find(`#${cancelId}`);
      cancelBtn.on("click", cancel);
      const timeTag = navbar.find(`#${timeTagId}`);
      return setCountdown(timeTag, deadline);
    }
  );
  closeBtn.on("click", () => {
    for (const disposer of countdownDisposers) {
      disposer();
    }
    closeBtn.off("click");
    addTooltip(clearHints(target), originalSubject);
  });
};

const compareDeadlines = (a: Hint, b: Hint): number =>
  (a.relay.upcoming?.deadline || 0) - (b.relay.upcoming?.deadline || 0);

export const showHints = ({
  hints,
  navbar,
}: {
  hints: Hint[];
  navbar: JQuery;
}) => {
  const hasMultipleTasks =
    hints.filter(({ relay }) => hasUpcomingTask(relay)).length > 1;
  hints.sort(compareDeadlines);
  let title = "";
  let content: string[] = [];
  let target: JQuery | undefined = undefined;
  let originalSubject = "";
  const items: PopoverItem[] = [];
  for (const { key, relay, control } of hints) {
    const isRelayHavingTask = hasUpcomingTask(relay);
    if (!isRelayHavingTask || target) {
      addTooltip(control, relay.label_text);
    }
    if (!isRelayHavingTask) {
      continue;
    }
    const { upcoming, label_text: subject } = relay;
    const dateObj = new Date(upcoming.deadline);
    const dateISO = dateObj.toISOString();
    const dateLocalized = dateObj.toLocaleString();
    const timeLeft = formatDeadline(upcoming.deadline);
    const targetState = upcoming.target ? "ON" : "OFF";
    const [cancelId, timeTagId] = ["cancel-btn", "time-tag"].map(
      (prefix) => `${prefix}-${key}`
    );
    items.push({
      cancelId,
      timeTagId,
      deadline: upcoming.deadline,
      cancel: () => cancelTask(key, upcoming),
    });
    target = target || control;
    originalSubject = originalSubject || subject;
    const upcomingHTML = `${subject} goes <span class="label">${targetState}</span>`;
    const timeHTML = `<time id="${timeTagId}" datetime="${dateISO}" title="${dateLocalized}">${timeLeft}</time>`;
    const cancelHTML = `<button id="${cancelId}" class="btn btn-mini" type="button">Cancel</button>`;
    title = hasMultipleTasks
      ? `<span>Several relay switches ahead</span>${closeBtnHTML}`
      : `<span>${upcomingHTML}</span>${closeBtnHTML}`;
    content.push(
      hasMultipleTasks
        ? `<div><span>${upcomingHTML} ${timeHTML}</span>${cancelHTML}</div>`
        : `<div>${timeHTML}${cancelHTML}</div>`
    );
  }
  if (target) {
    addPopover({
      target,
      navbar,
      originalSubject,
      title,
      content,
      items,
    });
  }
};
