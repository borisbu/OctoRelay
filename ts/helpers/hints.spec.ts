import MockDate from "mockdate";
import { countdownMock, disposerMock } from "../mocks/countdown";
import { addTooltip, clearHints, addPopover } from "./hints";

describe("Hints helpers", () => {
  const elementMock: Record<
    "popover" | "tooltip" | "find" | "on" | "off",
    jest.Mock
  > = {
    popover: jest.fn(() => elementMock),
    tooltip: jest.fn(() => elementMock),
    find: jest.fn(() => elementMock),
    on: jest.fn(() => elementMock),
    off: jest.fn(() => elementMock),
  };
  const jQueryMock = jest.fn((subject: string | (() => void)) => {
    if (typeof subject === "function") {
      return subject();
    }
    return elementMock;
  });

  Object.assign(global, {
    $: jQueryMock,
  });

  beforeAll(() => {
    MockDate.set("2023-08-13T22:30:00");
  });

  afterEach(() => {
    elementMock.popover.mockClear();
    elementMock.tooltip.mockClear();
    elementMock.find.mockClear();
    elementMock.on.mockClear();
    elementMock.off.mockClear();
  });

  afterAll(() => {
    MockDate.reset();
  });

  describe("clearHints() helper", () => {
    test("should remove both tooltip and popover from the target element", () => {
      clearHints(elementMock as unknown as JQuery);
      expect(elementMock.popover).toHaveBeenCalledWith("destroy");
      expect(elementMock.tooltip).toHaveBeenCalledWith("destroy");
    });
  });

  describe("addTooltip() helper", () => {
    test("should add tooltip to the target element", () => {
      addTooltip(elementMock as unknown as JQuery, "test");
      expect(elementMock.tooltip).toHaveBeenCalledWith({
        placement: "bottom",
        title: "test",
      });
    });
  });

  describe("addPopover() helper", () => {
    test("should add a popover to the target element", async () => {
      const cancelMock = jest.fn();
      addPopover({
        target: elementMock as unknown as JQuery,
        title: "Title",
        content: ["line1", "line2"],
        navbar: elementMock as unknown as JQuery,
        originalSubject: "Relay",
        items: [
          {
            cancelId: "cancel-id",
            deadline: Date.now() + 60 * 1000,
            timeTagId: "time-id",
            cancel: cancelMock,
          },
        ],
      });
      expect(elementMock.popover.mock.calls).toMatchSnapshot(".popover()");
      expect(elementMock.find.mock.calls).toMatchSnapshot(".find()");
      expect(countdownMock).toHaveBeenCalledWith(elementMock, 1691965860000);
      expect(elementMock.on).toHaveBeenCalledTimes(2); // cancel, close
      expect(elementMock.on.mock.calls[0][1]).toEqual(cancelMock);
      const close = elementMock.on.mock.calls[1][1];
      expect(elementMock.off).not.toHaveBeenCalled();
      expect(disposerMock).not.toHaveBeenCalled();
      close();
      expect(elementMock.off).toHaveBeenCalledWith("click");
      expect(disposerMock).toHaveBeenCalledTimes(1);
      expect(elementMock.tooltip).toHaveBeenCalledWith({
        placement: "bottom",
        title: "Relay",
      });
    });
  });
});
