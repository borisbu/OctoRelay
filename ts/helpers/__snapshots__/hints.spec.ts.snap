// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hints helpers addPopover() helper should add a popover to the target element: .find() 1`] = `
[
  [
    "#pop-closer",
  ],
  [
    "#cancel-id",
  ],
  [
    "#time-id",
  ],
]
`;

exports[`Hints helpers addPopover() helper should add a popover to the target element: .popover() 1`] = `
[
  [
    {
      "animation": false,
      "content": "line1line2",
      "html": true,
      "placement": "bottom",
      "title": "Title",
      "trigger": "manual",
    },
  ],
  [
    "show",
  ],
]
`;
